packages/backend/src/models/CatalogModel/CatalogModel.ts

```diff
@@ -607,4 +610,66 @@ export class CatalogModel {
             catalogTagsMigrateIn,
         );
     }
+
+    async getCatalogItemsWithIcons(projectUuid: string) {
+        let query = this.database(CatalogTableName)
+            .column(
+                `${CatalogTableName}.catalog_search_uuid`,
+                `${CatalogTableName}.cached_explore_uuid`,
+                `${CatalogTableName}.project_uuid`,
+                `${CatalogTableName}.name`,
+                `${CatalogTableName}.type`,
+                `${CatalogTableName}.field_type`,
+                `${CatalogTableName}.icon`,
+                {
+                    explore_base_table: this.database.raw(
+                        `${CachedExploreTableName}.explore->>'baseTable'`,
+                    ),
+                },
+            )
+            .leftJoin(
+                CachedExploreTableName,
+                `${CatalogTableName}.cached_explore_uuid`,
+                `${CachedExploreTableName}.cached_explore_uuid`,
+            );
+
+        query = query
+            .where(`${CatalogTableName}.project_uuid`, projectUuid)
+            .whereNotNull(`${CatalogTableName}.icon`)
+            .groupBy(
+                `${CatalogTableName}.catalog_search_uuid`,
+                `${CatalogTableName}.cached_explore_uuid`,
+                `${CatalogTableName}.project_uuid`,
+                `${CatalogTableName}.name`,
+                `${CatalogTableName}.type`,
+                `${CatalogTableName}.field_type`,
+                `explore_base_table`,
+            );
+
+        const itemsWithIcons: (DbCatalog & {
+            explore_base_table: string;
+        })[] = await query;
+
+        return itemsWithIcons.map<CatalogItemsWithIcons>((i) => ({
+            catalogSearchUuid: i.catalog_search_uuid,
+            cachedExploreUuid: i.cached_explore_uuid,
+            projectUuid: i.project_uuid,
+            name: i.name,
+            type: i.type,
+            fieldType: i.field_type,
+            exploreBaseTable: i.explore_base_table,
+            icon: i.icon,
+        }));
+    }
+
+    async updateCatalogItemIcon(
+        catalogSearchUuid: string,
+        icon: CatalogItemIcon | null,
+    ): Promise<void> {
+        await this.database(CatalogTableName)
+            .where('catalog_search_uuid', catalogSearchUuid)
+            .update({
+                icon,
+            });
+    }
```

Comment:
This should perhaps receive a list of updates and do everything in the same transaction instead of looping on the service.